api/src/recon.rs

```diff
@@ -0,0 +1,170 @@
+
+    async fn keys_since_highwater_mark(
+        &self,
+        highwater: HighwaterMark,
+        limit: usize,
+    ) -> Result<(ResumeToken, Vec<Self::Key>)>;
+}
+
+#[async_trait::async_trait]
+pub trait AccessModelStore: Clone + Send + Sync {
+    type Key: recon::Key;
+
+    async fn keys_since_highwater_mark(
+        &self,
+        highwater: HighwaterMark,
+        limit: i64,
+    ) -> anyhow::Result<(String, Vec<Self::Key>)>;
+}
+
+#[derive(Clone, Debug)]
+pub struct HighwaterMark(i64);
+
+impl TryFrom<&ResumeToken> for HighwaterMark {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &ResumeToken) -> Result<Self, Self::Error> {
+        Ok(Self(value.try_into()?))
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Wrapper for the recon client that provides additional functionality that isn't explicitly part of the recon protocol.
+pub struct ReconModelService<K, H, M> {
+    model: M,
+    recon: recon::Client<K, H>,
+}
+
+impl<K, H, M> ReconModelService<K, H, M>
+where
+    M: AccessModelStore<Key = K>,
+{
+    pub fn new(model: M, recon: recon::Client<K, H>) -> Self {
+        Self { model, recon }
+    }
+}
+
+#[async_trait]
+impl<K, H> ReconInterest for recon::Client<K, H>
+where
+    K: Key,
+    H: AssociativeHash + std::fmt::Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    type Key = K;
+    type Hash = H;
+
+    async fn insert(&self, key: Self::Key) -> Result<()> {
+        let _ = recon::Client::insert(self, key, None).await?;
+        Ok(())
+    }
+
+    async fn range(
+        &self,
+        start: Self::Key,
+        end: Self::Key,
+        offset: usize,
+        limit: usize,
+    ) -> Result<Vec<Self::Key>> {
+        Ok(recon::Client::range(self, start, end, offset, limit)
+            .await?
+            .collect())
+    }
+}
+
+#[async_trait]
+impl<K, H, M> crate::ReconModel for ReconModelService<K, H, M>
+where
+    K: Key,
+    H: AssociativeHash + std::fmt::Debug + Serialize + for<'de> Deserialize<'de>,
+    M: AccessModelStore<Key = K>,
+{
+    type Key = K;
+    type Hash = H;
+
+    async fn insert(&self, key: Self::Key, value: Option<Vec<u8>>) -> Result<()> {
+        let _ = recon::Client::insert(&self.recon, key, value).await?;
+        Ok(())
+    }
+
+    async fn range_with_values(
+        &self,
+        start: Self::Key,
+        end: Self::Key,
+        offset: usize,
+        limit: usize,
+    ) -> Result<Vec<(Self::Key, Vec<u8>)>> {
+        Ok(
+            recon::Client::range_with_values(&self.recon, start, end, offset, limit)
+                .await?
+                .collect(),
+        )
+    }
+    async fn value_for_key(&self, key: Self::Key) -> Result<Option<Vec<u8>>> {
+        recon::Client::value_for_key(&self.recon, key).await
+    }
+
+    async fn keys_since_highwater_mark(
+        &self,
+        highwater: HighwaterMark,
+        limit: usize,
+    ) -> Result<(ResumeToken, Vec<Self::Key>)> {
+        let (hw, res) = self
+            .model
```

Comment:
Does the `model` type have the `insert` `range_with_values` and `value_for_key` methods? Could we implement this ReconModel trait using only `M`? Then we can stop using the Recon::Client.

Reply:
We discussed this Friday and it seems like it, so I'll be working on this today